crates/storage/src/exec.rs

```diff
@@ -135,4 +152,46 @@ macro_rules! inst_ops {
     }
 }
 
+/// Same as `inst_ops` but calls the methods on context db. The function name must exist
+/// in the database instance
+macro_rules! inst_ops_auto {
+    {
+        ($db_method: ident, $base:ident, $ctx:ident $(<$($tparam:ident: $tpconstr:tt),+>)?) {
+            $($iname:ident($($aname:ident: $aty:ty),*) => $ret:ty;)*
+        }
+    } => {
+        inst_ops_common!{
+            ($base, $ctx $(<$($tparam: $tpconstr),+>)?) {
+                $( $iname($($aname: $aty),*) => $ret; )*
+            }
+        }
+
+        paste::paste! {
+            impl $(<$($tparam: $tpconstr + Sync + Send + 'static),+>)? ShimTrait for Inner $(<$($tparam),+>)? {
+                $(
+                    fn [<$iname _blocking>] (&self, $($aname: $aty),*) -> DbResult<$ret> {
+                        self.ctx.db.$db_method().$iname($($aname),*)
+                    }
+
+                    fn [<$iname _chan>] (&self, pool: &threadpool::ThreadPool, $($aname: $aty),*) -> DbRecv<$ret> {
+                        let (resp_tx, resp_rx) = tokio::sync::oneshot::channel();
+                        let ctx = self.ctx.clone();
+
+                        pool.execute(move || {
+                            let res = ctx.db.$db_method().$iname($($aname),*);
```

Comment:
This is overly coupled to the patterns that are downstream of the current structure of the `Database` trait and how it's been used in these `Context` types.  This is a problem since we're planning on removing the `Database` trait soon in favor of the specific database traits.

What would be *better* is to write a macro to instantiate the toplevel operation fns when they're just directly calling methods on a trait, where we could pass in an instance of `L2Database` directly, and then invoke that for each of these.  This would make it line up more with the existing macro's usage.

Reply:
Yes, agree that this is a bit coupled. I did not fully get what you mean afterwards though.

Another way I can think of is to have a `db()` method in the context that returns the appropriate db. And use `$ctx.db().$iname(..)` inside the macro. This makes it much less coupled and we don't have to pass the db method name in macro invocation. It would be ideal to enforce this via some trait on context but that seems tricky. Now it looks like this:

```rust
impl<D: BroadcastDatabase + Sync + Send + 'static> Context<D> {
   // ...

    // Add this method
    pub fn db(&self) -> &Arc<impl L1BroadcastDatabase> {
        self.db.l1_broadcast_db()
    }
}

inst_ops_auto! {
    // Now this is cleaner as we don't need to pass method name
    (BroadcastDbOps, Context<D: BroadcastDatabase>) {
        get_tx_entry(idx: u64) => Option<L1TxEntry>;
}
```

Corresponding change inisde macro def:

```rust
                   fn [<$iname _blocking>] (&self, $($aname: $aty),*) -> DbResult<$ret> {
                    // HERE                        
                    self.ctx.db().$iname($($aname),*)
                    }

                    fn [<$iname _chan>] (&self, pool: &threadpool::ThreadPool, $($aname: $aty),*) -> DbRecv<$ret> {
                        let (resp_tx, resp_rx) = tokio::sync::oneshot::channel();
                        let ctx = self.ctx.clone();

                        pool.execute(move || {
                            // AND HERE
                            let res = ctx.db().$iname($($aname),*);
                            if resp_tx.send(res).is_err() {
                                warn!("failed to send response");
                            }
                        });

                        resp_rx
                    }
```

Reply:
Also, I think that it will be cleaner if we expect the ops interface to implement the trait methods only and everything else be done on the Manager level.

Reply:
> I did not fully get what you mean afterwards though.

If it's being annoying, just make the `Context` type be generic over the `L1BroadcastDatabase` (or whatever else).

> I think that it will be cleaner if we expect the ops interface to implement the trait methods only 

Can you expand some more on what this would look like?

The goal I had in mind with this ops interface was to abstract over the low-level database provider trait and present the basic operations it has as both async and blocking varieties and without the extra effort involved in passing generic parameters around.  There's some space in there for indirection in case we need to introduce it but it's meant to be a thin wrapper.

The more authentically Rustic way to do that *would* be to make a proc macro that we use on the trait but that's pretty extra and it doesn't have to be that complicated.

Reply:
> Can you expand some more on what this would look like?

I think we are totally on the same page regarding the goals of the Ops interface. It should abstract over the provider trait. That's what I am trying to say, if we are doing just that then we can have those ops functions automatically generated as done in this PR. And yes, ideally via a proc macro on the trait/struct itself. Actually I even thought of pursuing that path, but it would be too much for now.

> > I think that it will be cleaner if we expect the ops interface to implement the trait methods only

So, by this I mean the ops helper functions can be automatically generated because we wouldn't be doing anything extra over there and having something like `inst_ops_auto!` that reduces a lot of boilerplate.

> If it's being annoying, just make the Context type be generic over the L1BroadcastDatabase (or whatever else).

Exposing a `db()` method in the `Context` type should be enough leaving the rest as is. Either way we need to expose `db()` method.

Reply:
Another strength of the current approach is that it doesn't make any assumptions about the structure of the types involved, you don't have to have any fields/methods with specific names.

We're debating new design directions with this now and if you had just done it in the established way then we would have moved on by now.

If the problem is the boilerplate then write another macro that wraps the `inst_ops!` invocation and encompasses the *whole* thing, I can do that and get back to you.

Reply:
Something like

```rs

macro_rules! inst_ops_simple {
    {
        ($base:ident, $ctx:ident <$tparam:ident: $tpconstr:tt>) {
            $($iname:ident($($aname:ident: $aty:ty),*) => $ret:ty;)*
        }
    } => {
        pub struct Context<$tparam : $tpconstr> {
            db: Arc<$tparam>,
        }

        impl<$tparam : $tpconstr + Sync + Send + 'static> Context<$tparam> {
            pub fn new(db: Arc<$tparam>) -> Self {
                Self { db }
            }

            pub fn into_ops(self, pool: threadpool::ThreadPool) -> $base {
                $base::new(pool, Arc::new(self))
            }
        }

        inst_ops!($base, Context<$tparam>) {
            $($iname ($($aname : $aty ),*) => $ret ;)*
        }

        $(
            fn $iname < $tparam : $tpconstr > (context: &Context<$tparam>, $($aname : $aty),* ) -> DbResult<$ret> {
                self.db.as_ref(). $iname ( $($aname),* )
            }
        )*
    }
}
```

Reply:
That might also let you get rid of a layer of `Arc` there too.